### PR TITLE
Base model

### PR DESCRIPTION
**Pull request by kninad**
Initial baseline model with the usual set of:
- data loading
- model
- loss functions
- training and eval loop (as scripts)

Currently the `Unet_fancy` model needs a bugfix so that we can set the model arch dynamically.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (Connor-Bowley) [inline .vscode/settings.json:1]: In general,  I don't recommend putting editor specific, settings in a code repository.

Some people like to put files like this in a separate repo that helps them bootstrap on a new machine. I don't really have a location I keep settings like this, I just adjust as needed on each machine.

File: eval.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+from torchvision.ops import sigmoid_focal_loss
+
+
+def eval(data_loader, model, exp_dir):
+    model.eval()
+    test_image_savepath = os.path.join(workspace.get_dir(exp_dir, workspace.evaluation_subdir), "images")
+    if not os.path.isdir(test_image_savepath):
+        os.makedirs(test_image_savepath)
+    pbar = tqdm(data_loader)
+    for _, data in enumerate(pbar):
```
Comment (Connor-Bowley) [inline eval.py]: It seems odd that you would use `enumerate` but ignore the enumeration

File: eval.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        img = data['image'].cuda()
+        # lab = data['mask'].cuda()
```
Comment (Connor-Bowley) [inline eval.py]: Recommend removing commented code unless it is serving some purpose.

File: eval.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+
+def main(args):
+    np.random.seed(2020)
+    torch.manual_seed(2020)
+    torch.cuda.manual_seed(2020)
+
+    if not os.path.isdir(args.exp_dir):
+        print(f"[ERROR] Experiment dir {args.exp_dir} does not exist!")
+        sys.exit(0)
```
Comment (Connor-Bowley) [inline eval.py]: In general, I would recommend `return 0` and `return main(args)` for the call to main. It makes it so this function could better be called by other code.

File: eval.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    # channels = (1, 64, 128, 256, 512, 1024)
+    # model = Unet2D(**specs["NetworkSpecs"])
+    channels = 1
+    classes = 1
+    model = Unet2D(channels, num_class=classes)
+    model = model.cuda()
+    workspace.load_model_checkpoint(experiment_dir, checkpoint, model)
+
+    trn_img_dir = os.path.join(train_data_dir, "images")
+    trn_lab_dir = os.path.join(train_data_dir, "labels")
```
Comment (Connor-Bowley) [inline eval.py]: Personal nitpick, when I see "lab" the first thing I think is some kind of laboratory, not labels. Not asking for a change, just something to consider.

**Comment by Connor-Bowley** [comment]
A couple things that should be added at this point (feel free to check the box when complete):

- [x] Add a [python requirement file](https://learnpython.com/blog/python-requirements-file/) (`requirements.txt` or something similar) that can be used to install the libraries you are using
- [x] Update the README such that someone given the data could run your code and interpret the output. This would include required folder structures, command line arguments, and description of the outputs. See the [github markdown syntax](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) for to see out to do things like [properly quote code](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code)
- [x] Add newlines to the ends of all your files that don't have them (the red circle with a dash in mark in when looking at the file in github).

File: models/blocks_unet.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
+from mimetypes import init
+from turtle import forward
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
```
Comment (Connor-Bowley) [inline models/blocks_unet.py]: Personal nitpick, F is too ambiguous a name. Something like `Func` would be far more descriptive.

File: models/blocks_unet.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+
+class ConvBlock(nn.Module):
+    def __init__(self, input_channels: int, output_channels: int) -> None:
+        super().__init__()
+        self.double_conv_block = nn.Sequential(
```
Comment (Connor-Bowley) [inline models/blocks_unet.py:11]: Pythonic syntax requested that "private" variables start with an underscore. This looks like it is probably private. See https://peps.python.org/pep-0008/#descriptive-naming-styles

File: models/unet.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        # Decoder Part: Go reverse along the collected features
+        xtmp = features[-1]  # initial value (encoder's last feature)
+        for i, d_blk in enumerate(self.dec_blocks):
+            idx = len(features) - 1 - i
+            xtmp = d_blk(xtmp, features[idx-1])  # note idx - 1
+        logits = self.output_conv(xtmp)
+        return logits
+
+
+class Unet2D_simple(nn.Module):
```
Comment (Connor-Bowley) [inline models/unet.py:38]: Recommend removing all the commented out code in this class for clarity

File: trainer.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+
+def train(train_loader, model, optimizer, loss_fn, epochs, exp_dir, save_every, writer):
+    for epoch in tqdm(range(1, epochs+1)):
+        model.train(True)
+        running_loss = 0.
+        dice_loss = 0.
+        focal_loss = 0.
+        pbar = tqdm(train_loader)
+        for _, data in enumerate(pbar):
```
Comment (Connor-Bowley) [inline trainer.py]: Another instance of enumerate while ignoring the enumeration

File: trainer.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+def main(args):
+
+    np.random.seed(2020)
+    torch.manual_seed(2020)
+    torch.cuda.manual_seed(2020)
+
+    if not os.path.isdir(args.exp_dir):
+        print(f"[ERROR] Experiment dir {args.exp_dir} does not exist!")
+        sys.exit(0)
```
Comment (Connor-Bowley) [inline trainer.py]: See other comment on `return 0` and `return main(args)`

File: trainer.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+    # channels = (1, 64, 128, 256, 512, 1024)
+    channels = 1
+    classes = 1
+    model = Unet2D(channels, num_class=classes)
+    # model = Unet2D(**specs["NetworkSpecs"])
+    model = model.cuda()
+    params = model.parameters()
+    optimizer = torch.optim.Adam(params, learning_rate)
+    # criterion = torch.nn.BCEWithLogitsLoss()
```
Comment (Connor-Bowley) [inline trainer.py]: Would it make sense to make criterion an enumeration command line argument?

File: utils/data.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+from pathlib import Path
+
+import numpy as np
+from PIL import Image
+import torch
+from torch.utils.data import DataLoader, RandomSampler, Subset
+from torchvision import transforms
+
+
+def get_rsz_transform(resize: int):
```
Comment (Connor-Bowley) [inline utils/data.py]: Recommend renaming to `get_resize_transform`. It was not initially clear to me what `rsz` meant.

File: utils/data.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    def __init__(self, images_dir: str, masks_dir: str, resize=None, load_in_ram=True):
+        self.images_dir = Path(images_dir)
+        self.masks_dir = Path(masks_dir)
+        self.resize = resize
+        self.load_in_ram = load_in_ram
+        self.transform = get_rsz_transform(
+            resize) if self.resize else get_basic_transform()
+        self.ids = [splitext(file)[0] for file in listdir(images_dir)]
+        self.ids.sort() # FOR consistency
+        if not self.ids:  # or not self.mask_ids:
```
Comment (Connor-Bowley) [inline utils/data.py]: This comment make me wonder if the `if` statement is correct or if it is missing a piece.

File: utils/loss.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+def dice_loss(pred, mask):
+    weit = 1 + 5*torch.abs(F.avg_pool2d(mask,kernel_size=31,stride=1,padding=15)-mask)
+    pred = torch.sigmoid(pred)
+    inter = ((pred * mask) * weit).sum(dim=(2,3))
+    union = ((pred + mask) * weit).sum(dim=(2,3))
+    wdiss = 1 - (2*inter + 0.5)/(union+0.5)
+    return wdiss.mean()
+
+class DiceLoss(nn.Module):
```
Comment (Connor-Bowley) [inline utils/loss.py:25]: It feels a little strange to me that `dice_loss` and `DiceLoss` in the same file have no obvious relation.

**Review by Connor-Bowley — APPROVED** [review]

File: .vscode/settings.json (hunk 1 above)
Comment (kninad) [reply .vscode/settings.json:1]: Sure, that can actually go into my dotfiles repo.

File: eval.py (hunk 4 above)
Comment (kninad) [reply eval.py]: I don't foresee the use of main()  since trainer.py and eval.py can be thought of main programs themselves calling using the different functions and classes. But I agree that return(0) is a graceful way to exit the main function.

File: models/blocks_unet.py (hunk 6 above)
Comment (kninad) [reply models/blocks_unet.py]: Yes, I agree! Using `F` for `torch.nn.functional` is standard practice but perhaps for our code, it can lead to name conflicts.

File: utils/data.py (hunk 13 above)
Comment (kninad) [reply utils/data.py]: Are you referring to the `if not self.ids:` statement?

File: utils/loss.py (hunk 14 above)
Comment (kninad) [reply utils/loss.py:25]: Yeah one is implemented as a function and the other as a class. Its better to stick to one way of doing it (sticking to class).

File: trainer.py (hunk 11 above)
Comment (kninad) [reply trainer.py]: I think its better to take it via the specs.json file since the `criterion` or the loss function is like a hyper-parameter for the experiment run.

File: models/blocks_unet.py (hunk 6 above)
Comment (Connor-Bowley) [reply models/blocks_unet.py]: If using `F` is standard practice, then it may be worth keeping. I have not really used `torch` so I was not aware. Up to you.

File: trainer.py (hunk 11 above)
Comment (Connor-Bowley) [reply trainer.py]: Makes sense

File: utils/data.py (hunk 13 above)
Comment (Connor-Bowley) [reply utils/data.py]: I was referring the comment `# or not self.mask_ids` which you removed already so no further action needed.
